### PR TITLE
feat: implement function last_day

### LINKED ISSUE #821: built-in-function: Support `last_day(TIMESTAMP)` function
**Is your feature request related to a problem? Please describe.**

### last_day(date|timestamp|int64)
Returns the last day of the month to which the `date` belongs to.

Notices: We only need to support last_day(timestamp)

### Examples:
```sql
> SELECT last_day(timestamp("2020-05-22 10:43:40"));
 2020-05-31
> SELECT last_day(timestamp("2020-02-12 10:43:40"));
 2020-02-29
> SELECT last_day(date("2020-02-12"));
 2020-02-29
```

### Reference:
https://spark.apache.org/docs/latest/api/sql/#last_day
**Describe the solution you'd like**

We do provide a built-in function development guide for you. 
Please check [OpenMLDB Build-In Function Develop Guide](https://github.com/4paradigm/OpenMLDB/blob/main/docs/en/built_in_function_develop_guide.md) or the [Chinese version](https://github.com/4paradigm/OpenMLDB/blob/main/docs/cn/built_in_function_develop_guide.md) for help.

Please make sure you:
- [ ] Add built-in C++ function in src/udf/udf.h and src/udf/udf.cc
- [ ] Register function to default library from function `void DefaultUdfLibrary::InitTimeAndDateUdf()` in src/udf/default_udf_library.cc.
- [ ] Add related unit test in src/codegen/udf_ir_builder_test.cc
- [ ] Documenting function 

If you have any questions or concerns, please let us know. We appreciate every piece of code, comment.
**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by hezean**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature: implement built-in function `last_day`

* **What is the new behavior (if this is a feature change)?**

Closes #821

### CODE REVIEW

**Comment by hezean** [comment]
Sorry for my oversight that when I was coding I forgot "we only need to support last_day(timestamp)". Please decide whether the `int64` and `Date` version need to be removed, thanks.

File: hybridse/src/udf/default_udf_library.cc [hunk 1]
```diff
@@ -2109,6 +2109,63 @@ void DefaultUdfLibrary::InitTimeAndDateUdf() {
...
+            Example:
+            @code{.sql}
+                select last_day(timestamp("2020-05-22 10:43:40"));
+                -- output 2020-05-31
+                select last_day(timestamp("2020-02-12 10:43:40"));
+                -- output 2020-02-29
+                select last_day(timestamp("2021-02-12"));
+                -- output 2021-02-28
+            @endcode
+            @since 0.5.3
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: please update it to next unrelease version, now it will be 0.6.0

File: hybridse/src/codegen/udf_ir_builder_test.cc [hunk 2]
```diff
@@ -227,6 +227,18 @@ TEST_F(UdfIRBuilderTest, weekofyear_date_udf_test) {
         CheckUdf<int32_t, Date>("weekofyear", 22, date);
     }
 }
+TEST_F(UdfIRBuilderTest, last_day_date_udf_test) {
```
Comment (aceforeverd) [inline hybridse/src/codegen/udf_ir_builder_test.cc]: always use camel case for test name

File: hybridse/src/codegen/udf_ir_builder_test.cc [hunk 3]
```diff
@@ -257,6 +269,12 @@ TEST_F(UdfIRBuilderTest, weekofyear_timestamp_udf_test) {
     Timestamp time(1590115420000L);
     CheckUdf<int32_t, Timestamp>("weekofyear", 21, time);
 }
+TEST_F(UdfIRBuilderTest, last_day_timestamp_udf_test) {
+    CheckUdf<Nullable<Date>, Timestamp>("last_day", nullptr,
+                                        Timestamp(std::numeric_limits<int64_t>::max()));
```
Comment (aceforeverd) [inline hybridse/src/codegen/udf_ir_builder_test.cc]: fix the cpplint issue

File: hybridse/src/udf/udf.cc [hunk 4]
```diff
@@ -184,6 +184,51 @@ int32_t weekofyear(Date *date) {
...
+void last_day(int64_t ts, Date *output, bool *is_null) {
+    time_t time = (ts + TZ_OFFSET) / 1000;
+    struct tm t;
+    if (nullptr == gmtime_r(&time, &t)) {
+        *is_null = true;
+        return;
+    }
+    gmtime_r(&time, &t);
+    try {
+        boost::gregorian::date d = boost::gregorian::date_from_tm(t);
```
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: Since there is previously a Seg fault issue in https://github.com/4paradigm/OpenMLDB/pull/2159#issuecomment-1185081618 that was sadly pointed to the boost datetime, so I recommend implement it with abseil time instead.  Similar as I previously did in https://github.com/4paradigm/OpenMLDB/pull/2175. We hope boost can be removed in the future.

Tell me If you have any other thoughts :)

**Comment by aceforeverd** [comment]
> Sorry for my oversight that when I was coding I forgot "we only need to support last_day(timestamp)". Please decide whether the `int64` and `Date` version need to be removed, thanks.

I think it's fine to keep the three

File: hybridse/src/udf/udf.cc (hunk 4 above)
Comment (hezean) [reply hybridse/src/udf/udf.cc]: Thanks for your detailed instruction!

**Comment by hezean** [comment]
Hi @aceforeverd, all CI passed now, can you please review again?

**Comment by hezean** [comment]
Hi @aceforeverd, all CI passed now, can you please review again?

File: hybridse/src/udf/udf.cc [hunk 5]
```diff
@@ -184,6 +184,32 @@ int32_t weekofyear(Date *date) {
     }
 }
 
+void last_day(int64_t ts, Date *output) {
+    time_t time = (ts + TZ_OFFSET) / 1000;
+    struct tm t;
+    gmtime_r(&time, &t);
```
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: how about `absl::ToCivilDay(FromUnixSeconds(ts) ...)` ?

File: hybridse/src/codegen/udf_ir_builder_test.cc [hunk 6]
```diff
@@ -333,10 +351,20 @@ TEST_F(UdfIRBuilderTest, weekofyear_int64_udf_test) {
...
-TEST_F(UdfIRBuilderTest, inc_int32_udf_test) {
+TEST_F(UdfIRBuilderTest, LastdayInt64UdfTest) {
+    CheckUdf<Date, int64_t>("last_day", Date(2020, 05, 31),
+                            1589958000000L);  // 2020-05-22
+    CheckUdf<Date, int64_t>("last_day", Date(2022, 07, 31),
+                            1658966400000L);  // 2022-07-28
+    CheckUdf<Date, int64_t>("last_day", Date(2022, 02, 28),
+                            1644451200000L);  // 2022-02-10
+    CheckUdf<Date, int64_t>("last_day", Date(2020, 02, 29),
+                            1581292800000L);  // 2020-02-10
```
Comment (aceforeverd) [inline hybridse/src/codegen/udf_ir_builder_test.cc]: add a few case to test invalid inputs, they should output null

File: hybridse/src/udf/udf.cc [hunk 7]
```diff
@@ -184,6 +184,32 @@ int32_t weekofyear(Date *date) {
     }
 }
 
+void last_day(int64_t ts, Date *output) {
```
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: if ts < 0, output should be null

File: hybridse/src/udf/udf.cc [hunk 8]
```diff
@@ -184,6 +185,38 @@ int32_t weekofyear(Date *date) {
...
+    absl::CivilDay civil_day = absl::ToCivilDay(absl::FromUnixMillis(ts),
+                                                absl::FixedTimeZone(TZ_OFFSET / 1000));
+    absl::CivilMonth next_month = absl::CivilMonth(civil_day) + 1;
+    absl::CivilDay last_day = absl::CivilDay(next_month) - 1;
+    *output = Date(static_cast<int32_t>(last_day.year()), last_day.month(), last_day.day());
+    if (is_null) {
+        *is_null = false;
+    }
+}
+void last_day(Timestamp *ts, Date *output) { last_day(ts->ts_, output, nullptr); }
```
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: you can add a `bool*` parameter too, so it is no need to check `is_null == nullptr`

**Review by aceforeverd — COMMENTED** [review]
Others looks good to me

**Comment by aceforeverd** [comment]
@zhanghaohit help review this

File: hybridse/src/udf/default_udf_library.cc [hunk 9]
```diff
@@ -2109,6 +2109,40 @@ void DefaultUdfLibrary::InitTimeAndDateUdf() {
...
+            Example:
+            @code{.sql}
+                select last_day(timestamp("2020-05-22 10:43:40"));
+                -- output 2020-05-31
+                select last_day(timestamp("2020-02-12 10:43:40"));
+                -- output 2020-02-29
+                select last_day(timestamp("2021-02-12"));
+                -- output 2021-02-28
+            @endcode
+            @since 0.6.0
```
Comment (zhanghaohit) [inline hybridse/src/udf/default_udf_library.cc]: since 0.6.1?

File: hybridse/src/udf/udf.h [hunk 10]
```diff
@@ -224,6 +224,10 @@ int32_t weekofyear(int64_t ts);
 int32_t weekofyear(Timestamp *ts);
 int32_t weekofyear(Date *ts);
 
+void last_day(int64_t ts, Date *output, bool *is_null);
+void last_day(Timestamp *ts, Date *output);
+void last_day(Date *ts, Date *output, bool *is_null);
```
Comment (zhanghaohit) [inline hybridse/src/udf/udf.h]: possible to define as `const Date*` and `const Timestamp*`?

**Review by zhanghaohit — CHANGES_REQUESTED** [review]

File: hybridse/src/udf/udf.cc (hunk 8 above)
Comment (hezean) [reply hybridse/src/udf/udf.cc]: So the function should be registered as returning Nullable<Date>, but actually always returns a Date for not null Timestamp input?

File: hybridse/src/udf/udf.h (hunk 10 above)
Comment (hezean) [reply hybridse/src/udf/udf.h]: Surely that's a better practice. Should I also change other udfs' input pointer param as const? That could be more unified

File: hybridse/src/udf/default_udf_library.cc [hunk 11]
```diff
@@ -2109,6 +2109,40 @@ void DefaultUdfLibrary::InitTimeAndDateUdf() {
...
+        )";
+
+    RegisterExternal("last_day")
+            .args<int64_t>(reinterpret_cast<void*>(static_cast<void (*)(int64_t, Date*, bool*)>(v1::last_day)))
+            .return_by_arg(true)
+            .returns<Nullable<Date>>()
+            .doc(last_day_doc);
+
+    RegisterExternal("last_day")
+            .args<Timestamp>(reinterpret_cast<void*>(static_cast<void (*)(const Timestamp*, Date*, bool*)>(v1::last_day)))
```
Comment (zhanghaohit) [inline hybridse/src/udf/default_udf_library.cc]: lint error

**Review by zhanghaohit — CHANGES_REQUESTED** [review]

**Review by zhanghaohit — APPROVED** [review]
LGTM

File: hybridse/src/udf/udf.h (hunk 10 above)
Comment (zhanghaohit) [reply hybridse/src/udf/udf.h]: It's ok to just change the `const` for the newly-added functions. 

For other udfs' signature, it would be better you can create a new PR and do the fix, if you are interested. Thanks.

**Comment by hezean** [comment]
Hi @aceforeverd, can you help merge this pr? many thanks!

**Review by aceforeverd — APPROVED** [review]

**Comment by aceforeverd** [comment]
merged!
